src/highdicom/io.py

```diff
@@ -0,0 +1,780 @@
+
+    Parameters
+    ----------
+    fp: pydicom.filebase.DicomFile
+        Pointer for DICOM PS3.10 file stream positioned at the first byte of
+        the Pixel Data element following the empty Basic Offset Table (BOT)
+    number_of_frames: int
+        Total number of frames in the dataset
+
+    Returns
+    -------
+    List[int]
+        Offset of each Frame item in bytes from the first byte of the Pixel Data
+        element following the BOT item
+
+    Note
+    ----
+    Moves the pointer back to the first byte of the Pixel Data element
+    following the BOT item (the first byte of the first Frame item).
+
+    Raises
+    ------
+    IOError
+        When file pointer is not positioned at first byte of first Frame item
+        after Basic Offset Table item or when parsing of Frame item headers
+        fails
+    ValueError
+        When the number of offsets doesn't match the specified number of frames
+
+    """
+    initial_position = fp.tell()
+    offset_values = []
+    current_offset = 0
+    i = 0
+    while True:
+        frame_position = fp.tell()
+        tag = TupleTag(fp.read_tag())
+        if int(tag) == SequenceDelimiterTag:
+            break
+        if int(tag) != ItemTag:
+            fp.seek(initial_position, 0)
+            raise IOError(
+                'Building Basic Offset Table (BOT) failed. '
+                f'Expected tag of Frame item #{i} at position {frame_position}.'
+            )
+        length = fp.read_UL()
+        if length % 2:
+            fp.seek(initial_position, 0)
+            raise IOError(
+                'Building Basic Offset Table (BOT) failed. '
+                f'Length of Frame item #{i} is not a multiple of 2.'
+            )
+        elif length == 0:
+            fp.seek(initial_position, 0)
+            raise IOError(
+                'Building Basic Offset Table (BOT) failed. '
+                f'Length of Frame item #{i} is zero.'
+            )
+
+        first_two_bytes = fp.read(2, 1)
+        if not fp.is_little_endian:
+            first_two_bytes = first_two_bytes[::-1]
+
+        # In case of fragmentation, we only want to get the offsets to the
+        # first fragment of a given frame. We can identify those based on the
+        # JPEG and JPEG 2000 markers that should be found at the beginning and
+        # end of the compressed byte stream.
+        if first_two_bytes in _START_MARKERS:
+            current_offset = frame_position - initial_position
+            offset_values.append(current_offset)
+
+        i += 1
+        fp.seek(length - 2, 1)  # minus the first two bytes
+
+    if len(offset_values) != number_of_frames:
+        raise ValueError(
+            'Number of frame items does not match specified Number of Frames.'
+        )
+    else:
+        basic_offset_table = offset_values
+
+    fp.seek(initial_position, 0)
+    return basic_offset_table
+
+
+def _build_icc_transform(metadata: Dataset) -> Union[ImageCmsTransform, None]:
```

Comment:
Consider contributing to pydicom.pixel_data_handlers.util instead of here?

Reply:
I already created an [issue](https://github.com/pydicom/pydicom/issues/1244) and would be happy to contribute that to *pydicom*. However, this function requires *Pillow*, which is optional in *pydicom*. @scaramallion any suggestions for a potential PR?

Reply:
In that case, it would probably go into pydicom.pixel_data_handlers.pillow_handler